Auto-breakout - Module implementation

SaiTeja0510 opened the pull request:
##### SUMMARY
Auto-breakout - Module implementation

##### GitHub Issues

| GitHub Issue # |
| -------------- |
| N/A|


##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request


##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
sonic_system

##### OUTPUT
<!--- Paste the functionality test result below -->


##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below
N/A
```
<!--- Measure the code coverage before and after the change by running the UT and ensure that the "coverage after the change" is not less than the coverage "before the change". Note that the unit testing coverage can be manually executed using the pytest tool or ansible-test tool. -->

##### Checklist:

- [X] I have performed a self-review of my own code to ensure there are no formatting, linting, or security issues
- [X] I have verified that new and existing unit tests pass locally with my changes
- [X] I have not allowed coverage numbers to degenerate
- [] I have maintained at least 90% code coverage
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have maintained backward compatibility or have provided any relevant "breaking_changes" descriptions in a "fragment" file in the "changelogs/fragments" directory of this repository.

##### How Has This Been Tested?

- [X] Regression test HTML report:
[sonic_system_test_results.pdf](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/14372269/sonic_system_test_results.pdf)
- [X] Regression test detailed terminal log:
[system_autobreakout__UT.txt](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/14372275/system_autobreakout__UT.txt)
- [X] facts_UT log:
[system_autobreakout_facts_UT.txt](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/14372290/system_autobreakout_facts_UT.txt)

Edit 1:

Adding new UT logs which includes lates changes of system module.
[system_autobreakout_UT_2.txt](https://github.com/ansible-collections/dellemc.enterprise_sonic/files/15095460/system_autobreakout_UT_2.txt)

kerry-meyer commented:
Thank you for providing these changes to support the Auto-breakout configuration attribute.

In order to proceed with the code review for this PR, please do the following:

- Update the "module" file for the "system" resource module (plugins/modules/sonic_system.py) to document the new "auto_breakout" configuration attribute. (Only a sentence or two is needed.)
- Fix the (white space) sanity failures for the PR contents. REF: "Details" link for any failing CI/sanity check for the PR (near the bottom of the "Conversation" tab of the PR) or https://github.com/ansible-collections/dellemc.enterprise_sonic/actions/runs/8002707024/job/21856717555?pr=342
- Add UT test cases that will cause the new code to be executed and tested during UT execution. (Changes are needed in the following files to add this UT coverage: "tests/unit/modules/network/sonic/test_sonic_system.py" and "tests/unit/modules/network/sonic/fixtures/sonic_system.yaml".)

SaiTeja0510 commented:
> Thank you for providing these changes to support the Auto-breakout configuration attribute.
> 
> In order to proceed with the code review for this PR, please do the following:
> 
> * Update the "module" file for the "system" resource module (plugins/modules/sonic_system.py) to document the new "auto_breakout" configuration attribute. (Only a sentence or two is needed.)
> * Add a "fragment" file to describe this change in the "changelogs/fragments" file for use in the upcoming release notes. (See other files in the target directory for examples for format and content of this one-line file.)
> * Fix the (white space) sanity failures for the PR contents. REF: "Details" link for any failing CI/sanity check for the PR (near the bottom of the "Conversation" tab of the PR) or https://github.com/ansible-collections/dellemc.enterprise_sonic/actions/runs/8002707024/job/21856717555?pr=342
> * Add UT test cases that will cause the new code to be executed and tested during UT execution. (Changes are needed in the following files to add this UT coverage: "tests/unit/modules/network/sonic/test_sonic_system.py" and "tests/unit/modules/network/sonic/fixtures/sonic_system.yaml".)

Thank you for the feedback.
I have addressed the comments. Please review and let me know if anything else needs to me changed.

mingjunzhang2019 commented on plugins/module_utils/network/sonic/config/system/system.py:
```diff
@@ -395,6 +411,7 @@ def get_replaced_config(self, have, want):
 
         w_hostname = want.get('hostname', None)
         w_intf_name = want.get('interface_naming', None)
+        w_auto_breakout_mode = want.get('auto_breakout', None)
```
Should line#416 be added with "or  w_auto_breakout_mode is not None"?

SaiTeja0510 replied:
> Should line#416 be added with "or w_auto_breakout_mode is not None"?

Yes that had to be added. Addressed it. Thank you for pointing it out.

mingjunzhang2019 commented on plugins/modules/sonic_system.py:
```diff
@@ -75,6 +75,13 @@
+        type: str
+        choices:
+          - ENABLE
+          - DISABLE
```
The following "EXAMPLES" section should be added with some examples for auto_breakout, which should come from the resource model examples files.

SaiTeja0510 replied:
> The following "EXAMPLES" section should be added with some examples for auto_breakout, which should come from the resource model examples files.

Added examples in the file.

mingjunzhang2019 commented on plugins/module_utils/network/sonic/facts/system/system.py:
```diff
@@ -86,6 +86,22 @@ def get_anycast_addr(self):
+                auto_breakout_val = auto_breakout_data[0]['auto-breakout']
+                data = {'auto-breakout': auto_breakout_val}
+            else:
+                data = {}
```
If line#96 is false through, data will not have definition. So data can be initiated before line#96 and remove line#101 and 102.

SaiTeja0510 replied:
> If line#96 is false-through, the "data" will not have definition. The "data" can be initiated before line#96 and remove line#101 and 102.

Agreed. Updated the code.

kerry-meyer reviewed (approved):
The proposed code changes and posted test results look good.

Thank you for providing this enhancement to the functionality of the Dell enterprise_sonic Ansible resource module collection.

Approved.
